src/compose/ComposeBox.js

```diff
@@ -153,270 +136,352 @@ const updateTextInput = (textInput, text) => {
-
-  inputBlurTimeoutId: ?TimeoutID = null;
-
-  state = {
-    isMessageFocused: false,
-    isTopicFocused: false,
-    isFocused: false,
-    isMenuExpanded: false,
-    height: 20,
-    topic:
-      this.props.initialTopic
-      ?? (isTopicNarrow(this.props.narrow) ? topicOfNarrow(this.props.narrow) : ''),
-    message: this.props.initialMessage ?? '',
+  const messageInputRef = React.useRef<$FlowFixMe | null>(null);
+  const topicInputRef = React.useRef<$FlowFixMe | null>(null);
+
+  const mentionWarnings = React.useRef<React$ElementRef<typeof MentionWarnings> | null>(null);
+
+  const inputBlurTimeoutId = useRef<?TimeoutID>(null);
+
+  const [isMenuExpanded, setIsMenuExpanded] = useState<boolean>(false);
+  const [height, setHeight] = useState<number>(20);
+  const [numUploading, setNumUploading] = useState<number>(0);
+
+  const [focusState, setFocusState] = useState<{|
+    message: boolean,
+    topic: boolean,
+
+    /** Almost the same as message || topic ... except debounced, to stay
+     * true while those flip from false/true to true/false and back. */
+    either: boolean,
+  |}>({
+    message: false,
+    topic: false,
+    either: false,
+  });
+
+  // The topic input is currently uncontrolled, for performance concerns;
+  // see #2738. That means if you change this state, it won't cause the
+  // input value to change; callers should call updateTextInput for that.
+  // But if callers do that, they should *also* update this state, because
+  // it's our best read of what the actual value is, so it needs to be
+  // up-to-date. That's what setTopicInputValue (far below) is for; it does
+  // both.
+  // TODO(?): Make custom Hook to encapsulate this uncontrolled-input logic.
+  const [topicInputState, setTopicInputState] = useState<{| value: string |}>({
+    value: props.initialTopic ?? (isTopicNarrow(props.narrow) ? topicOfNarrow(props.narrow) : ''),
+  });
+
+  // The message input is currently uncontrolled, for performance concerns;
+  // see #2738. That means if you change this state, it won't cause the
+  // input value to change; callers should call updateTextInput for that.
+  // But if callers do that, they should *also* update this state, because
+  // it's our best read of what the actual value is, so it needs to be
+  // up-to-date. That's what setTopicInputValue (far below) is for; it does
+  // both.
+  // TODO(?): Make custom Hook to encapsulate this uncontrolled-input logic.
+  const [messageInputState, setMessageInputState] = useState<{|
+    value: string,
+    selection: InputSelection,
+  |}>({
+    value: props.initialMessage ?? '',
     selection: { start: 0, end: 0 },
-    numUploading: 0,
-  };
+  });
 
-  componentWillUnmount() {
-    clearTimeout(this.inputBlurTimeoutId);
-    this.inputBlurTimeoutId = null;
-  }
+  useEffect(
+    () => () => {
+      clearTimeout(inputBlurTimeoutId.current);
+      inputBlurTimeoutId.current = null;
+    },
+    [],
+  );
```

Comment:
All this blur-timeout logic, together with `focusState`, would be a great chunk of logic to move to a custom Hook as a followup.

(Probably I wouldn't try to make it any more general than it is -- it'd still use the names "message" and "topic", and live in this file. But all the hairy details would be in one place, separate from the other hairy details of this component.)